modules/indexing/src/test/java/org/apache/ignite/internal/metric/IndexPagesMetricsPageDisplacementTest.java

```diff
@@ -0,0 +1,189 @@
+
+    /** */
+    private IgniteEx grid;
+
+    /** */
+    private IgniteCache<Integer, Person> cache;
+
+    /** */
+    private IndexPageCounter idxPageCounter;
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        cleanPersistenceDir();
+
+        super.beforeTest();
+
+        grid = startGrid();
+        grid.cluster().state(ClusterState.ACTIVE);
+
+        CacheConfiguration<Integer, Person> cacheConfiguration = new CacheConfiguration<Integer, Person>(TEST_CACHE_NAME)
+            .setIndexedTypes(Integer.class, Person.class);
+        cache = grid.createCache(cacheConfiguration);
+
+        idxPageCounter = new IndexPageCounter(grid, true);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        grid.close();
+
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return super.getConfiguration(igniteInstanceName)
+            .setFailureHandler(new StopNodeFailureHandler())
+            .setDataStorageConfiguration(
+                new DataStorageConfiguration()
+                    .setDefaultDataRegionConfiguration(
+                        new DataRegionConfiguration()
+                            .setMaxSize(12 * 1024 * 1024) // 12 MB
+                            .setPersistenceEnabled(true)
+                            .setPageEvictionMode(DataPageEvictionMode.RANDOM_LRU)
+                            .setEvictionThreshold(0.1)
+                    )
+            );
+    }
+
+    /**
+     * Inserts data into the cache as long as it fills enough for some data pages to get dumped to the
+     * storage. Since index page metrics should reflect the number of in-memory pages, the metric value is expected to
+     * go down.
+     */
+    @Test
+    public void testPageDisplacement() throws IgniteCheckedException {
+        int grpId = grid.cachex(TEST_CACHE_NAME).context().groupId();
+
+        DataRegion dataRegion = grid.context().cache().context().database().dataRegion(null);
+        PageMetrics pageMetrics = dataRegion.metrics().cacheGrpPageMetrics(grpId);
+
+        int personId = 0;
+        long idxPagesOnDisk;
+        long idxPagesInMemory;
+
+        do {
+            // insert data into the cache until some index pages get displaced to the storage
+            for (int i = 0; i < 100; i++, personId++)
+                cache.put(personId, new Person(personId, "foobar"));
+
+            idxPagesOnDisk = getIdxPagesOnDisk(grpId).size();
+            idxPagesInMemory = idxPageCounter.countIdxPagesInMemory(grpId);
+
+            assertThat(pageMetrics.indexPages().value(), is(idxPagesInMemory));
+
+        } while (idxPagesOnDisk <= idxPagesInMemory);
+
+        // load pages back into memory and check that the metric value has increased
+        touchIdxPages(grpId);
```

Comment:
Is this ok that you don't release these pages?

Reply:
I have a comment inside the `touchIdxPages` method: `// acquire, but not release all pages, so that they don't get displaced back to the storage`. Or do you think my reasoning is wrong?